core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1153,154 +1101,207 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
+    preferredReplicasForTopicsByBrokers.foreach {
+      case(leaderBroker, topicAndPartitionsForBroker) => {
+        var imbalanceRatio: Double = 0
+        var topicsNotInPreferredReplica: Map[TopicAndPartition, Seq[Int]] = topicAndPartitionsForBroker
+          .filter { case (topicPartition, _) =>
+          controllerContext.partitionLeadershipInfo.contains(topicPartition) &&
+            controllerContext.partitionLeadershipInfo(topicPartition).leaderAndIsr.leader != leaderBroker
+        }
+        debug("topics not in preferred replica " + topicsNotInPreferredReplica)
+        val totalTopicPartitionsForBroker = topicAndPartitionsForBroker.size
+        val totalTopicPartitionsNotLedByBroker = topicsNotInPreferredReplica.size
+        imbalanceRatio = totalTopicPartitionsNotLedByBroker.toDouble / totalTopicPartitionsForBroker
+        trace("leader imbalance ratio for broker %d is %f".format(leaderBroker, imbalanceRatio))
+        // check ratio and if greater than desired ratio, trigger a rebalance for the topic partitions
+        // that need to be on this broker
+        if (imbalanceRatio > (config.leaderImbalancePerBrokerPercentage.toDouble / 100)) {
+          topicsNotInPreferredReplica.keys.foreach { topicPartition =>
+            // do this check only if the broker is live and there are no partitions being reassigned currently
+            // and preferred replica election is not in progress
+            if (controllerContext.liveBrokerIds.contains(leaderBroker) &&
+                controllerContext.partitionsBeingReassigned.isEmpty &&
+                controllerContext.partitionsUndergoingPreferredReplicaElection.isEmpty &&
+                !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic) &&
+                controllerContext.allTopics.contains(topicPartition.topic)) {
+              onPreferredReplicaElection(Set(topicPartition), true)
+            }
+          }
+        }
+      }
+    }
+  }
 
-    def handleStateChanged(state: KeeperState) {
-      // do nothing, since zkclient will do reconnect for us.
+  def getControllerID(): Int = {
+    controllerContext.zkUtils.readDataMaybeNull(ZkUtils.ControllerPath)._1 match {
+      case Some(controller) => KafkaController.parseControllerId(controller)
+      case None => -1
     }
+  }
 
-    /**
-     * Called after the zookeeper session has expired and a new session has been created. You would have to re-create
-     * any ephemeral nodes here.
-     *
-     * @throws Exception On any error.
-     */
-    @throws[Exception]
-    def handleNewSession() {
-      info("ZK expired; shut down all controller components and try to re-elect")
-      if (controllerElector.getControllerID() != config.brokerId) {
-        onControllerResignation()
-        inLock(controllerContext.controllerLock) {
-          controllerElector.elect
+  def addToControllerEventQueue(controllerEvent: ControllerEvent): Unit = {
```

Comment:
Thinking about it, I think you were in the right track when you were passing a queue to the various classes. It's just that it should have been a `ControllerQueue` instead of the underlying `LinkedBlockingQueue`. That way, we get:

1. The listeners have access to a much more limited interface instead of all of the controller. It seems like that's all they need?
2. We have an easy place to check who uses the Controller queue
3. We have a reasonable abstraction instead of this method that looks a bit out of place

Any reason why this would be bad?